proposals/22866-unified-queue-implementation.md

```diff
@@ -0,0 +1,80 @@
+# Unified queue implementation
+
+Goal:
+- Upcoming activities run in order as listed (one queue) and can be canceled.
+- So that I have the ability to prioritize an important action (ex. lock/wipe) by cancelling all upcoming activity.
+
+The following entities are part of the upcoming activities queue:
```

Comment:
They are! I didn't single them out in this list because profile delivery are via MDM commands (`InstallProfile` or `RemoveProfile`: https://github.com/fleetdm/fleet/blob/7ad3188f9c0baf3184c8552d27b7afd8d14df1b0/server/mdm/apple/commander.go#L48-L72).

So AIUI if we had that unified queue table, the "Install Profile" MDM commands would be queued there instead of queued directly to the nano/windows MDM tables, and when it's their time to run, they would be transferred to the nano/windows MDM table for processing via the MDM protocol.

Reply:
> That presents added complexity in the form of profile reconciliation (cron) and profile verification

I'm not sure I get the impact on those jobs? My understanding is that reconciliation would work as it does today, and would enqueue in the new queue instead of in the MDM tables? And profile verification would work the same too?

Reply:
I guess I'm really just expanding on a subset of "subtle issues in behavior" you mention [here](https://github.com/fleetdm/fleet/pull/23501/files#diff-f2cc7af549e313ccbf7a6122cbcdaf299c43bf20e0eccc532c86134a3b5ba521R23). There's a lot of business logic and assumptions about how things work like pending status, profile modifications, etc. baked into the service and datastore methods for those parts that we'll need to take extra care to get right.

Reply:
Dropping a quick note for our future selves to consider potential impacts the Puppet workflow (nothing springs to mind right now so we should just circle back when we're farther down the road with this).

Reply:
Good call-out, this is a bit of a blind spot to me.